app/main.py

```diff
@@ -1,44 +1,67 @@
+import asyncio
 import time
+from typing import Awaitable, Any
 
 from iot.devices import HueLightDevice, SmartSpeakerDevice, SmartToiletDevice
 from iot.message import Message, MessageType
 from iot.service import IOTService
 
 
-def main() -> None:
+async def run_sequence(*functions: Awaitable[Any]) -> None:
+    for function in functions:
+        await function
+
+
+async def run_parallel(*functions: Awaitable[Any]) -> None:
+    await asyncio.gather(*functions)
+
+
+async def main() -> None:
     # create an IOT service
     service = IOTService()
 
     # create and register a few devices
     hue_light = HueLightDevice()
     speaker = SmartSpeakerDevice()
     toilet = SmartToiletDevice()
-    hue_light_id = service.register_device(hue_light)
-    speaker_id = service.register_device(speaker)
-    toilet_id = service.register_device(toilet)
+    devices = [hue_light, speaker, toilet]
 
     # create a few programs
-    wake_up_program = [
-        Message(hue_light_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up"),
-    ]
-
-    sleep_program = [
-        Message(hue_light_id, MessageType.SWITCH_OFF),
-        Message(speaker_id, MessageType.SWITCH_OFF),
-        Message(toilet_id, MessageType.FLUSH),
-        Message(toilet_id, MessageType.CLEAN),
-    ]
+    devices = await asyncio.gather(
+        *[service.register_device(device) for device in devices]
+    )
+    hue_light_id, speaker_id, toilet_id = devices[0], devices[1], devices[2]
 
-    # run the programs
-    service.run_program(wake_up_program)
-    service.run_program(sleep_program)
+    await run_parallel(
+        service.run_program([Message(hue_light_id, MessageType.SWITCH_ON)]),
+        run_sequence(
+            service.run_program(
+                [
+                    Message(hue_light_id, MessageType.SWITCH_ON),
+                    Message(speaker_id,
+                            MessageType.PLAY_SONG,
+                            "Rick Astley - Never Gonna Give You Up"),
+                ]
+            ),
+            service.run_program(
+                [
+                    Message(hue_light_id, MessageType.SWITCH_OFF),
+                    Message(speaker_id, MessageType.SWITCH_OFF)
+                ]
+            ),
+            service.run_program(
+                [
+                    Message(toilet_id, MessageType.FLUSH),
+                    Message(toilet_id, MessageType.CLEAN),
+                ]
+            )
+        )
+    )
```

Comment:
The `run_parallel` function is used to execute tasks concurrently, which is appropriate here. However, ensure that all functions passed to `run_parallel` are asynchronous and return awaitable objects. Otherwise, this might lead to unexpected behavior.